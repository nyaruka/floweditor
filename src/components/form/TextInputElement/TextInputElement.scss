@import '../../../variables.scss';

.count {
    display: -webkit-box;
    display: -ms-flexbox;
    display: flex;
    -webkit-box-pack: end;
        -ms-flex-pack: end;
            justify-content: flex-end;
    margin: 5px 0;
}

.tooltip {
    position: relative;
    display: inline-block;
    font-size: 15px;
    margin-left: 5px;
}

.tooltip .tooltiptext {
    visibility: hidden;
    display: -webkit-box;
    display: -ms-flexbox;
    display: flex;
    -webkit-box-orient: vertical;
    -webkit-box-direction: normal;
        -ms-flex-direction: column;
            flex-direction: column;
    -webkit-box-pack: justify;
        -ms-flex-pack: justify;
            justify-content: space-between;
    height: 220px;
    width: 175px;
    background-color: white;
    color: $grey;
    text-align: left;
    padding: 12px;
    border-radius: 4px;
    font-size: 14px;
    line-height: 20px;
    -webkit-box-shadow: $shadow_down;
    box-shadow: $shadow_down;

    /* Position tooltip text */
    position: absolute;
    z-index: 1;
    bottom: 138%;
    left: -490%;
    margin-left: -60px;

    /* Fade in tooltip */
    opacity: 0;
    -webkit-transition: opacity 0.25s;
    -o-transition: opacity 0.25s;
    transition: opacity 0.25s;
}

/* Show tooltip text when mouse hovers over the tooltip container */
.tooltip:hover .tooltiptext {
    visibility: visible;
    opacity: 1;
}

.tooltiprow {
    display: -webkit-box;
    display: -ms-flexbox;
    display: flex;
    -webkit-box-pack: justify;
        -ms-flex-pack: justify;
            justify-content: space-between;
}

.option_list {
    overflow-y: auto;
    list-style-type: none;
    padding: 0;
    max-height: 300px;
    margin-bottom: 0px;
}

.option_description {
    font-size: 12px;
    padding: 3px 0;
    font-weight: 300;
}

.option {
    border-top: 1px solid transparent;
    border-bottom: 1px solid transparent;
    padding: 5px;
    font-size: 14px;
    &.selected {
        border-color: $blue_antd;
        background: $blue_antd;
        color: $text_color;
        font-weight: 500;
        &.first_option {
            border-top-color: transparent !important;
        }
    }
}

.completion_container {
    position: absolute;
    background: #fff;
    margin-top: 20px;
    width: 250px;
    border-radius: 4px;
    -webkit-box-shadow: $shadow_down;
            box-shadow: $shadow_down;
    z-index: $z_completion;
}

.wrapper {
    position: relative;
}

.help {
    font-size: 10px;
    color: #999;
    padding: 4px;
    border-radius: 0 0 4px 4px;
    background-color: #f5f5f5;
}

.hidden {
    display: none;
}

.textinput {
    border-radius: 4px;
    width: 100%;
    color: $grey;
    font-size: 14px;
    font-weight: 300;
    border: 1px solid $light_grey;
    overflow: none;
    outline: none;
    -webkit-box-shadow: none;
    box-shadow: none;
    padding: 10px 10px;
    -webkit-box-sizing: border-box;
    box-sizing: border-box;
    height: 37px;
    &::-webkit-input-placeholder {
        color: $placeholder_color;
    }
    &::-moz-placeholder {
        color: $placeholder_color;
    }
    &:-ms-input-placeholder {
        color: $placeholder_color;
    }
    &:-moz-placeholder {
        color: $placeholder_color;
    }
}

textarea.textinput {
    height: 120px;
    padding: 10px;
    overflow: auto;
}
