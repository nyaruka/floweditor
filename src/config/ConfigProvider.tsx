import * as React from 'react';
import { shape, string, objectOf } from 'prop-types';
import { Languages, Endpoints, FlowEditorConfig } from '../flowTypes';
import { Language } from '../component/LanguageSelector';

interface ConfigProviderProps {
    config: FlowEditorConfig;
    children: any;
}

export interface ConfigProviderContext {
    endpoints: Endpoints;
    languages: Languages;
}

// Prop-type definitions (required by React's context API: https://reactjs.org/docs/context.html)
export const assetHostPT = string;
export const endpointsPT = shape({
    fields: string,
    groups: string,
    engine: string,
    contacts: string,
    flows: string,
    activity: string
});
export const languagesPT = objectOf(string);
// ----------------------------------------------------------------------------------------------

export const SINGLE_CHILD_ERROR = 'ConfigProvider expects only one child component.';
export const VALID_CHILD_ERROR =
    'ConfigProvider expects a valid React element: https://reactjs.org/docs/react-api.html#isvalidelement';

export default class ConfigProvider extends React.Component<ConfigProviderProps> {
    public static childContextTypes = {
        assetHost: assetHostPT,
        endpoints: endpointsPT,
        languages: languagesPT
    };

    constructor(props: ConfigProviderProps) {
        super(props);

        if (React.Children.count(props.children) > 1) {
            throw new Error(SINGLE_CHILD_ERROR);
        } else if (!React.isValidElement(props.children)) {
            throw new Error(VALID_CHILD_ERROR);
        }
    }

    public getChildContext(): ConfigProviderContext {
        return {
            languages: this.props.config.languages,
            endpoints: this.props.config.endpoints
        };
    }

    public render(): JSX.Element {
        return React.Children.only(this.props.children);
    }
}
