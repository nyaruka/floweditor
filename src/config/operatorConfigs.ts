export interface Operator {
    type: string;
    verboseName: string;
    operands: number;
    // Default category name to use if specified
    categoryName?: string;
}

export interface OperatorMap {
    [propName: string]: Operator;
}

export const operatorConfigList: Operator[] = [
    { type: 'has_any_word', verboseName: 'has any of the words', operands: 1 },
    { type: 'has_all_words', verboseName: 'has all of the words', operands: 1 },
    { type: 'has_phrase', verboseName: 'has the phrase', operands: 1 },
    { type: 'has_only_phrase', verboseName: 'has only the phrase', operands: 1 },
    { type: 'has_beginning', verboseName: 'starts with', operands: 1 },
    {
        type: 'has_text',
        verboseName: 'has some text',
        operands: 0,
        categoryName: 'Has Text'
    },
    {
        type: 'has_number',
        verboseName: 'has a number',
        operands: 0,
        categoryName: 'Has Number'
    },
    {
        type: 'has_number_between',
        verboseName: 'has a number between',
        operands: 2
    },
    { type: 'has_number_lt', verboseName: 'has a number below', operands: 1 },
    {
        type: 'has_number_lte',
        verboseName: 'has a number at or below',
        operands: 1
    },
    { type: 'has_number_eq', verboseName: 'has a number equal to', operands: 1 },
    {
        type: 'has_number_gte',
        verboseName: 'has a number at or above',
        operands: 1
    },
    { type: 'has_number_gt', verboseName: 'has a number above', operands: 1 },
    { type: 'has_date', verboseName: 'has a date', operands: 0, categoryName: 'Has Date' },
    { type: 'has_date_lt', verboseName: 'has a date before', operands: 1 },
    { type: 'has_date_eq', verboseName: 'has a date equal to', operands: 1 },
    { type: 'has_date_gt', verboseName: 'has a date after', operands: 1 },
    { type: 'has_run_status', verboseName: 'has a run status of', operands: 1 },
    { type: 'has_group', verboseName: 'is in the group', operands: 1 },
    {
        type: 'has_phone',
        verboseName: 'has a phone number',
        operands: 0,
        categoryName: 'Has Phone'
    },
    {
        type: 'has_email',
        verboseName: 'has an email',
        operands: 0,
        categoryName: 'Has Email'
    },
    {
        type: 'has_error',
        verboseName: 'has an error',
        operands: 0,
        categoryName: 'Has Error'
    },
    {
        type: 'has_value',
        verboseName: 'is not empty',
        operands: 0,
        categoryName: 'Not Empty'
    }
];

export const operatorConfigMap: OperatorMap = operatorConfigList.reduce(
    (map: OperatorMap, operatorConfig: Operator) => {
        map[operatorConfig.type] = operatorConfig;
        return map;
    },
    {}
);

export type GetOperatorConfig = (type: string) => Operator;

/**
 * Shortcut for constant lookup of operator config in operator configs map
 * @param {string} type - The type of the operator config to return, e.g. 'send_msg'
 * @returns {Object} - The operator config found at operatorConfigs[type] or -1
 */
export const getOperatorConfig = (type: string): Operator => operatorConfigMap[type];
