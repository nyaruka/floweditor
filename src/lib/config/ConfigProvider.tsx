import * as React from 'react';
import { shape, string, objectOf } from 'prop-types';
import { Languages, Endpoints } from '../flowTypes';
import { Language } from '../component/LanguageSelector';

export interface ConfigProviderContext {
    assetHost: string;
    endpoints: Endpoints;
    languages: Languages;
}

// Prop-type definitions (required by React's context API)
export const assetHostPT = string;
export const endpointsPT = shape({
    fields: string,
    groups: string,
    engine: string,
    contacts: string,
    flows: string,
    activity: string
});
export const languagesPT = objectOf(string);
// -------------------------------------------------------

export default class ConfigProvider extends React.Component<any> {
    public static childContextTypes = {
        assetHost: assetHostPT,
        endpoints: endpointsPT,
        languages: languagesPT
    };

    constructor(props: any) {
        super(props);

        if (React.Children.count(props.children) > 1) {
            throw new Error('ConfigProvider expects only one child component.');
        }
    }

    public getChildContext(): ConfigProviderContext {
        return {
            assetHost: this.props.assetHost,
            languages: this.props.languages,
            endpoints: this.props.endpoints
        };
    }

    public render(): JSX.Element {
        return React.Children.only(this.props.children);
    }
}
