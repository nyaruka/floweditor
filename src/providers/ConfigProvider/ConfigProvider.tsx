import * as React from 'react';
import { ConfigProviderContext } from './configContext';
import { childContextPT } from './propTypes';
import configContext from './configContext';

export default class ConfigProvider extends React.Component<any> {
    public static childContextTypes = childContextPT;

    constructor(props: any) {
        super(props);

        if (React.Children.count(props.children) > 1) {
            throw new Error('ConfigProvider expects to receive only one child component.');
        }
    }

    public getChildContext(): ConfigProviderContext {
        return configContext;
    }

    public render(): JSX.Element {
        return React.Children.only(this.props.children);
    }
}
