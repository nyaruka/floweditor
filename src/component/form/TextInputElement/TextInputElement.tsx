import * as React from 'react';
import * as classNames from 'classnames/bind';
import { react as bindCallbacks } from 'auto-bind';
import getCaretCoordinates from 'textarea-caret';
import setCaretPosition from 'get-input-selection';
import ComponentMap, { CompletionOption } from '../../../services/ComponentMap';
import CharCount from './CharCount';
import FormElement, { FormElementProps } from '../FormElement';
import { Type } from '../../../config';
import { UnicodeCharMap, getOptionsList, getMsgStats, filterOptions, isValidURL } from './helpers';
import {
    KEY_P,
    KEY_UP,
    KEY_N,
    KEY_DOWN,
    KEY_AT,
    KEY_ESC,
    KEY_TAB,
    KEY_ENTER,
    KEY_BACKSPACE,
    KEY_SPACE,
    COMPLETION_HELP
} from './constants';
import * as styles from './TextInputElement.scss';
import * as shared from '../FormElement.scss';

export enum Count {
    SMS = 'SMS'
}

export interface Coordinates {
    left: number;
    top: number;
}

export interface HTMLTextElement {
    value: string;
    selectionStart: number;
    selectionEnd: number;
    focus(): void;
}

interface TextInputProps extends FormElementProps {
    value: string;
    ComponentMap: ComponentMap;
    config?: Type;
    __className?: string;
    count?: Count;
    url?: boolean;
    textarea?: boolean;
    placeholder?: string;
    autocomplete?: boolean;
    focus?: boolean;
    showInvalid?: boolean;
    onChange?(event: React.ChangeEvent<HTMLTextElement>): void;
    onBlur?(event: React.ChangeEvent<HTMLTextElement>): void;
}

export interface TextInputState {
    value: string;
    errors: string[];
    caretOffset: number;
    caretCoordinates: Coordinates;
    completionVisible: boolean;
    selectedOptionIndex: number;
    matches: CompletionOption[];
    query: string;
    options: CompletionOption[];
    parts?: string[];
    characterCount?: number;
    unicodeChars?: UnicodeCharMap;
}

type InitialState = Pick<
    TextInputState,
    | 'caretOffset'
    | 'caretCoordinates'
    | 'errors'
    | 'completionVisible'
    | 'selectedOptionIndex'
    | 'matches'
    | 'query'
>;

const initialState: InitialState = {
    caretOffset: 0,
    caretCoordinates: { left: 0, top: 0 },
    errors: [],
    completionVisible: false,
    selectedOptionIndex: 0,
    matches: [],
    query: ''
};

const cx = classNames.bind({ ...styles, ...shared });

export default class TextInputElement extends React.Component<TextInputProps, TextInputState> {
    private selectedEl: HTMLLIElement;
    private textEl: HTMLTextElement;

    constructor(props: TextInputProps) {
        super(props);

        this.state = {
            value: this.props.value,
            options: getOptionsList(this.props.autocomplete, this.props.ComponentMap),
            ...initialState,
            ...(this.props.count && this.props.count === Count.SMS
                ? getMsgStats(this.props.value)
                : {})
        };

        bindCallbacks(this, {
            include: [/^on/, /Ref$/, 'setSelection', 'validate']
        });
    }

    private selectedElRef(ref: HTMLLIElement): HTMLLIElement {
        return (this.selectedEl = ref);
    }

    private textElRef(ref: HTMLTextElement): HTMLTextElement {
        return (this.textEl = ref);
    }

    public componentWillReceiveProps(nextProps: TextInputProps): void {
        if (nextProps.value !== this.props.value) {
            this.setState({ value: nextProps.value });
        }
    }

    public componentDidMount(): void {
        return this.props.focus && this.focusInput();
    }

    public componentDidUpdate(previous: TextInputProps): void {
        return this.selectedEl && this.selectedEl.scrollIntoView(false);
    }

    private onKeyDown(event: React.KeyboardEvent<HTMLTextElement>): void {
        if (!this.props.autocomplete) {
            return;
        }

        switch (event.keyCode) {
            case KEY_P:
                if (!event.ctrlKey) {
                    break;
                }
            case KEY_UP:
                if (this.state.completionVisible) {
                    this.setSelection(this.state.selectedOptionIndex - 1);
                    event.preventDefault();
                }
                break;
            case KEY_N:
                if (!event.ctrlKey) {
                    break;
                }
            case KEY_DOWN:
                if (this.state.completionVisible) {
                    this.setSelection(this.state.selectedOptionIndex + 1);
                    event.preventDefault();
                }
                break;
            case KEY_AT:
                this.setState({
                    completionVisible: true,
                    caretCoordinates: getCaretCoordinates(this.textEl, this.textEl.selectionEnd)
                });
                break;
            case KEY_ESC:
                if (this.state.completionVisible) {
                    this.setState({
                        completionVisible: false
                    });
                    event.preventDefault();
                    event.stopPropagation();
                }
                break;
            case KEY_TAB:
            case KEY_ENTER:
                if (this.state.completionVisible && this.state.matches.length > 0) {
                    var option = this.state.matches[this.state.selectedOptionIndex];
                    var newValue = this.state.value.substr(
                        0,
                        this.state.caretOffset - this.state.query.length
                    );
                    newValue += option.name;

                    var newCaret = newValue.length;
                    newValue += this.state.value.substr(this.state.caretOffset);

                    var query = '';
                    var completionVisible = false;
                    var matches: CompletionOption[] = [];
                    if (event.keyCode === KEY_TAB) {
                        query = option.name;
                        matches = filterOptions(this.state.options, query);
                        completionVisible = matches.length > 0;
                    }

                    this.setState(
                        {
                            query,
                            value: newValue,
                            matches,
                            caretOffset: newCaret,
                            completionVisible,
                            selectedOptionIndex: 0
                        },
                        () => {
                            setCaretPosition(this.textEl, newCaret);
                        }
                    );

                    /** TODO: set caret position */
                    event.preventDefault();
                    event.stopPropagation();
                }
                break;
            case KEY_BACKSPACE:
                // Iterate backwards on our value until we reach either a space or @
                var caret = event.currentTarget.selectionStart - 1;
                for (let i = caret - 1; i >= 0; i--) {
                    const curr = this.state.value[i];
                    /** Space, don't do anything but break out */
                    if (curr === ' ') {
                        break;
                    }

                    // if '@' we display completion menu again
                    if (curr === '@') {
                        query = this.state.value.substr(i + 1, caret - i - 1);
                        matches = filterOptions(this.state.options, query);
                        completionVisible = matches.length > 0;
                        return this.setState({
                            query,
                            matches,
                            caretOffset: caret,
                            completionVisible,
                            selectedOptionIndex: 0,
                            caretCoordinates: getCaretCoordinates(this.textEl, i)
                        });
                    }
                }

                // We are visible still but really shouldn't be, clear out
                if (this.state.completionVisible) {
                    this.setState({
                        query: '',
                        matches: [],
                        caretOffset: caret,
                        completionVisible: false,
                        selectedOptionIndex: 0
                    });
                }
                break;
            case KEY_SPACE:
                this.setState({
                    completionVisible: false
                });
                break;
        }
    }

    private onBlur(event: React.ChangeEvent<HTMLTextElement>): void {
        this.setState(
            {
                query: '',
                matches: [],
                caretOffset: 0,
                selectedOptionIndex: 0,
                completionVisible: false
            },
            () => this.props.onBlur && this.props.onBlur(event)
        );
    }

    private onChange(event: React.ChangeEvent<HTMLTextElement>): void {
        const { currentTarget: { value, selectionStart } } = event;

        const updates: Partial<TextInputState> = {
            value
        };

        if (this.props.autocomplete) {
            if (this.state.completionVisible) {
                let query = value.substring(0, selectionStart);
                const lastIdx = query.lastIndexOf('@');

                if (lastIdx > -1) {
                    query = query.substring(lastIdx + 1);
                }

                updates.query = query;
                updates.matches = filterOptions(this.state.options, query);
            } else {
                if (this.props.count === Count.SMS) {
                    const stats = getMsgStats(value, true);

                    updates.parts = stats.parts;
                    updates.characterCount = stats.characterCount;
                    updates.unicodeChars = stats.unicodeChars;
                }
            }

            updates.caretOffset = selectionStart;
            updates.selectedOptionIndex = 0;
        }

        this.setState(updates as TextInputState);

        if (this.props.onChange) {
            this.props.onChange(event);
        }
    }

    private setSelection(selectedIdx: number): void {
        let selectedOptionIndex: number = selectedIdx;

        // Can't exceed the last option
        if (selectedIdx >= this.state.matches.length) {
            selectedOptionIndex = this.state.matches.length - 1;
        }

        // Can't go beyond the first option
        if (selectedIdx < 0) {
            selectedOptionIndex = 0;
        }

        if (selectedOptionIndex !== this.state.selectedOptionIndex) {
            this.setState({ selectedOptionIndex });
        }
    }

    public validate(): boolean {
        const errors: string[] = [];

        if (this.props.required) {
            if (!this.state.value) {
                errors.push(`${this.props.name} is required`);
            }
        }

        this.setState({ errors });

        // See if it should be a valid url
        if (errors.length === 0) {
            if (this.props.url) {
                if (!isValidURL(this.state.value)) {
                    errors.push('Enter a valid URL');
                }
            }
        }

        return errors.length === 0;
    }

    private focusInput(): void {
        const { value: { length } } = this.textEl;
        this.textEl.focus();
        this.textEl.selectionStart = length;
    }

    private getOption({ name, description }: CompletionOption, selected: boolean): JSX.Element {
        const optionName = <div data-spec="option-name">{name}</div>;

        if (selected) {
            return (
                <React.Fragment>
                    {optionName}
                    <div data-spec="option-desc" className={styles.option_description}>
                        {description}
                    </div>
                </React.Fragment>
            );
        }

        return optionName;
    }

    private getOptions(): JSX.Element[] {
        return this.state.matches.map((option: CompletionOption, idx: number) => {
            const optionClasses = [styles.option];

            if (idx === this.state.selectedOptionIndex) {
                optionClasses.push(styles.selected);

                if (idx === 0) {
                    optionClasses.push(styles.first_option);
                }

                return (
                    <li
                        ref={this.selectedElRef}
                        className={optionClasses.join(' ')}
                        key={option.name}>
                        {this.getOption(option, true)}
                    </li>
                );
            }

            return (
                <li className={optionClasses.join(' ')} key={option.name}>
                    {this.getOption(option, false)}
                </li>
            );
        });
    }

    public render(): JSX.Element {
        const textElClasses = cx({
            [styles.textinput]: true,
            [shared.invalid]: this.state.errors.length > 0 || this.props.showInvalid === true
        });

        const completionClasses = cx({
            [styles.completion_container]: true,
            [styles.hidden]: !this.state.completionVisible || this.state.matches.length === 0
        });

        const options: JSX.Element[] = this.getOptions();

        const charCount: JSX.Element =
            this.props.count && this.props.count === Count.SMS ? (
                <CharCount
                    count={this.state.characterCount}
                    parts={this.state.parts.length}
                    unicodeChars={this.state.unicodeChars}
                />
            ) : null;

        const sendMsgError =
            this.state.errors.length > 0 &&
            this.props.name === 'Message' &&
            this.props.config.type === 'send_msg';

        // Make sure we're rendering the right text element
        const TextElement = this.props.textarea ? 'textarea' : ('input' as string);

        const inputType = this.props.textarea ? undefined : 'text';

        return (
            <FormElement
                __className={this.props.__className}
                name={this.props.name}
                helpText={this.props.helpText}
                showLabel={this.props.showLabel}
                errors={this.state.errors}
                sendMsgError={sendMsgError}>
                <div className={styles.wrapper}>
                    <TextElement
                        data-spec="input"
                        ref={this.textElRef}
                        type={inputType}
                        className={textElClasses}
                        value={this.state.value}
                        onChange={this.onChange}
                        onBlur={this.onBlur}
                        onKeyDown={this.onKeyDown}
                        placeholder={this.props.placeholder}
                    />
                    <div
                        className={completionClasses}
                        style={this.state.caretCoordinates}
                        data-spec="completion-options">
                        <ul className={styles.option_list} data-spec="completion-list">
                            {options}
                        </ul>
                        <div className={styles.help} data-spec="completion-help">
                            {COMPLETION_HELP}
                        </div>
                    </div>
                </div>
                {charCount}
            </FormElement>
        );
    }
}
