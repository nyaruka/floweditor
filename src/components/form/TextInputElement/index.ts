import TextInputElement, { Count, CharacterSet, HTMLTextElement } from './TextInputElement';

export { Count, CharacterSet, HTMLTextElement };

export default TextInputElement;
