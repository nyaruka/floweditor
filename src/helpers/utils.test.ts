import { CharacterSet } from '../components/form/TextInputElement';
import { operatorConfigList } from '../providers/ConfigProvider/operatorConfigs';
import {
    addCommas,
    snakify,
    toBoolMap,
    validUUID,
    titleCase,
    getSelectClass,
    reorderList,
    jsonEqual,
    hasErrorType
} from './utils';

describe('utils >', () => {
    describe('toBoolMap >', () => {
        it('should turn a string array into a bool map', () => {
            const strArr: string[] = [
                'language',
                'facebook',
                'telegram',
                'email',
                'mailto',
                'name',
                'first name',
                'phone',
                'groups',
                'uuid',
                'created by',
                'modified by',
                'org',
                'is',
                'has',
                'tel'
            ];
            const expected: { [key: string]: boolean } = {
                'created by': true,
                email: true,
                facebook: true,
                'first name': true,
                groups: true,
                has: true,
                is: true,
                language: true,
                mailto: true,
                'modified by': true,
                name: true,
                org: true,
                phone: true,
                tel: true,
                telegram: true,
                uuid: true
            };

            expect(toBoolMap(strArr)).toEqual(expected);
        });
    });

    describe('addCommas >', () => {
        it('should insert commas into numbers where appropriate', () => {
            expect(addCommas(999)).toBe('999');
            expect(addCommas(10000)).toBe('10,000');
            expect(addCommas(10000000)).toBe('10,000,000');
        });
    });

    describe('snakify >', () => {
        it('should replace spaces with underscores', () => {
            expect(snakify('my result name')).toBe('my_result_name');
        });
    });

    describe('validUUID >', () => {
        it('matches a valid v4 UUID', () => {
            const v4 = '9ddd2483-3071-498d-bd4e-0fe9c0b2fa94';
            const v4Upper = '06FDAF8D-9905-4DA7-B94D-7BA3532D3953';

            expect(validUUID(v4)).toBeTruthy();
            expect(validUUID(v4Upper)).toBeTruthy();
        });

        it("doesn't match an invalid v4 UUID", () => {
            const v1 = '63955E30-B5DE-11E7-8F1A-0800200C9A66';
            const randStr = 'ZUBikEKpzoD7XOdo74Ux';

            expect(validUUID(v1)).toBeFalsy();
            expect(validUUID(randStr)).toBeFalsy();
        });
    });

    describe('getSelectClass >', () => {
        it('should return an empty array if passed "errors" arg less than 1', () =>
            expect(getSelectClass(0)).toEqual(''));

        it("should return an array containing react-select's invalid class", () =>
            expect(getSelectClass(1)).toEqual('select-invalid'));
    });

    describe('titleCase >', () =>
        it('should apply title case to each word in a given string', () =>
            ['one', 'one two', 'one two three', 'one,two', 'one, two'].forEach(str =>
                expect(titleCase(str)).toMatchSnapshot()
            )));

    describe('reorderList >', () =>
        it('should reorder a given list according to index params', () =>
            expect(reorderList([1, 2, 3], 2, 0)).toEqual([3, 1, 2])));

    describe('jsonEqual >', () => {
        const [anyWordOperator, allWordsOperator] = operatorConfigList;

        it('should return true if basic objects are equal in contents and order', () =>
            expect(jsonEqual(anyWordOperator, anyWordOperator)).toBeTruthy());

        it('should return false if basic objects are not equal in contents and order', () =>
            expect(jsonEqual(anyWordOperator, allWordsOperator)).toBeFalsy());
    });

    describe('hasErrorType >', () => {
        const errors = ['A category name is required.'];

        it('should return false if passed an empty error list', () =>
            expect(hasErrorType([], ['argument'])).toBeFalsy());

        it('should return true if query exits in a string in the error list', () =>
            expect(hasErrorType(errors, ['category'])).toBeTruthy());
    });
});
