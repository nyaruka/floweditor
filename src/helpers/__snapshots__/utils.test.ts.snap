// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils > titleCase > should apply title case to each word in a given string 1`] = `"One"`;

exports[`utils > titleCase > should apply title case to each word in a given string 2`] = `"One Two"`;

exports[`utils > titleCase > should apply title case to each word in a given string 3`] = `"One Two Three"`;

exports[`utils > titleCase > should apply title case to each word in a given string 4`] = `"One,Two"`;

exports[`utils > titleCase > should apply title case to each word in a given string 5`] = `"One, Two"`;
